### PR TITLE
add serial port class:

### PR DESCRIPTION
**Pull request by BitBander**
add serial port class for serial communication, and if it is necessary,
there is a relative communication protocol to apply.

### CODE REVIEW

File: common/include/serial_port.h [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+class SerialPort: public AbstractDevice {
+public:
+	enum character_size_type {character_size_5 = 5, character_size_6 = 6, \
+		character_size_7 = 7, character_size_8 = 8};
+	enum parity_type {parity_none = 0, parity_odd = 1, parity_even = 2};
+	enum stop_bits_type {stop_bits_one, stop_bits_onepointfive, stop_bits_two};
+	enum flow_control_type {flow_control_none, flow_control_software, flow_control_hardware};
+
+    static const constexpr char* kSerialPortName 	= "serial_port_name";
```
Comment (luxianzi) [inline common/include/serial_port.h]: Please pay attention to white spaces, you're mixing spaces and tabs, we are required to use tab only. Normally we should turn on the "show whitespace characters", it's a good habit.

File: common/include/abstract_protocol.h [hunk 2]
```diff
@@ -0,0 +1,27 @@
+/*
+ * abstract_protocol.h
+ *
+ *  Created on: 2020��9��30��
+ *      Author: duanning
+ */
+
+#ifndef INCLUDE_ABSTRACT_PROTOCOL_H_
```
Comment (luxianzi) [inline common/include/abstract_protocol.h]: Please remove all the path strings in re-include prevention macro definitions, e.g. `INCLUDE_`

File: common/include/abstract_protocol.h [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+class AbstractProtocol {
+public:
+
+	virtual int PackFrame() {};
+	virtual bool UnpackFrame() {};
+	virtual bool IsFrameComplete() {};
+	virtual bool IsFrameValid() {};
+
+private:
+	virtual int CalcCRC(vector<char>);
```
Comment (luxianzi) [inline common/include/abstract_protocol.h]: An "abstract" class should not have "private" members, if we do need it, please make it public. Private virtual member functions will not be reloaded.

File: common/include/serial_port.h [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    static const constexpr char* kSerialPortData = "serial_port_data";
+
+    SerialPort(const string& name);
```
Comment (luxianzi) [inline common/include/serial_port.h]: Please remove the parameters in constructors. Considering we are only using empty constructor parameter for device classes.

File: common/include/serial_port.h [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+				const vector<uint8_t>& value);
+	error_condition Get(const string& key, unsigned int& result);
+	error_condition Get(const string& key, \
+				vector<uint8_t>& result);
+
+private:
+	io_service io_s_;
+	serial_port serial_port_;
+	AbstractProtocol application_protocol_;
+	boost::system::error_code ec;
```
Comment (luxianzi) [inline common/include/serial_port.h]: This name of private member doesn't end with underscore.

File: common/source/lidar_protocol.cpp [hunk 6]
```diff
@@ -0,0 +1,32 @@
+/*
```
Comment (luxianzi) [inline common/source/lidar_protocol.cpp]: IMHO this file could be removed, we are not using it.

File: common/include/lidar_protocol.h [hunk 7]
```diff
@@ -0,0 +1,28 @@
+/*
```
Comment (luxianzi) [inline common/include/lidar_protocol.h]: IMHO this file could be removed, we are not using it. Please remove files which are not referenced in a commit.

File: common/include/serial_port.h [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+	struct {
```
Comment (luxianzi) [inline common/include/serial_port.h]: Please consider using "key pairs" to set parameter. E.g. we could call SerialPortDevice::Set("boardrate", 115200).

File: common/include/serial_port.h [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+#include <boost/asio.hpp>
+#include <boost/bind.hpp>
+#include "abstract_device.h"
+#include "abstract_protocol.h"
+
+using namespace boost::asio;
+
+class SerialPort: public AbstractDevice {
+public:
+	enum character_size_type {character_size_5 = 5, character_size_6 = 6, \
```
Comment (luxianzi) [inline common/include/serial_port.h]: Please remove these enums and consider using strings to specify a key name.

File: common/source/serial_port.cpp [hunk 10]
```diff
@@ -0,0 +1,124 @@
+/*
+ * serial_port.cpp
+ *
+ *  Created on: 2020��9��18��
+ *      Author: duanning
+ */
+
+#include "serial_port.h"
+
+SerialPort::SerialPort(const string& name):
```
Comment (luxianzi) [inline common/source/serial_port.cpp]: Same as above, please using empty parameter.

File: common/source/serial_port.cpp [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+	serial_params.kByteCommCharacterSize = character_size;
+	serial_params.kByteCommParity = parity;
+	serial_params.kByteCommStopBits = stop_bits;
+	serial_params.kByteCommFlowControl = flow_control;
+
+	if (ReConfig(config) != kNoError)
+		return;
+}
+
+SerialPort::~SerialPort() {
```
Comment (luxianzi) [inline common/source/serial_port.cpp:15]: What's the behavior when you destroy a `serial_port` object? Will it be closed?

File: common/source/serial_port.cpp [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+
+}
+
+error_condition SerialPort::ReConfig(const CommonKeyPairs &config) {
+	ClearInitialized();
+	auto serial_port_name = config.find(kSerialPortName);
```
Comment (luxianzi) [inline common/source/serial_port.cpp]: As I sugguested, to use string key name, we need to also change this function. When calling `Set` function, we could append a key pair to a private config member.

File: common/source/serial_port.cpp [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+	CommonKeyPairs config = {
+			{"hid_name", name},
+	};
+
+	ReConfig(config);
+}
+
+SerialPort::SerialPort(const string& name, AbstractProtocol& protocol):
+	serial_port_(io_s_),
+	application_protocol_(protocol) {
```
Comment (luxianzi) [inline common/source/serial_port.cpp]: `application_protocol_` is initialized but never referenced. IMHO the protocol is used when `Set` and `Get` data. At least we should pack or unpack the packet.

File: common/source/serial_port.cpp [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+	CommonKeyPairs config = {
+			{"hid_name", name},
```
Comment (luxianzi) [inline common/source/serial_port.cpp]: Why here is an `hid_name` key?

**Review by luxianzi — CHANGES_REQUESTED** [review]

File: common/include/serial_port.h (hunk 1 above)
Comment (BitBander) [reply common/include/serial_port.h]: Thank you. I will pay attention to this problem.

File: common/include/abstract_protocol.h (hunk 2 above)
Comment (BitBander) [reply common/include/abstract_protocol.h]: So, it should be:
#ifndef ABSTRACT_PROTOCOL_H_
?

File: common/include/abstract_protocol.h (hunk 3 above)
Comment (BitBander) [reply common/include/abstract_protocol.h]: Got it

File: common/include/abstract_protocol.h (hunk 2 above)
Comment (luxianzi) [reply common/include/abstract_protocol.h]: Yes

File: common/include/abstract_protocol.h (hunk 3 above)
Comment (luxianzi) [reply common/include/abstract_protocol.h]: Please make this function public and use `uint8_t` to replace `char`.

File: common/include/serial_port.h [hunk 15]
```diff
@@ -0,0 +1,70 @@
+/*
+ * serial_port.h
+ *
+ *  Created on: 2020��9��18��
+ *      Author: duanning
+ */
+
+#ifndef INCLUDE_SERIAL_PORT_H_
```
Comment (luxianzi) [inline common/include/serial_port.h]: Same issue as above.

File: common/include/serial_port.h [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+	enum parity_type {parity_none, parity_odd, parity_even};
+	enum stop_bits_type {stop_bits_one, stop_bits_onepointfive, stop_bits_two};
+	enum flow_control_type {flow_control_none, flow_control_software, flow_control_hardware};
+
+	static const constexpr char* kSerialPortName = "serial_port_name";
+	static const constexpr char* kSerialPortData = "serial_port_data";
+
+	SerialPort(void);
+	virtual ~SerialPort();
+	error_condition Init(const string& name,\
```
Comment (luxianzi) [inline common/include/serial_port.h]: Please consider using ReConfig.

File: common/include/serial_port.h [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+	error_condition Set(const string& key, \
+				const vector<uint8_t>& value);
+	error_condition Get(const string& key, unsigned int& result);
+	error_condition Get(const string& key, \
+				vector<uint8_t>& result);
+
+private:
+	io_service io_s_;
+	serial_port serial_port_;
+	AbstractProtocol application_protocol_;
```
Comment (luxianzi) [inline common/include/serial_port.h]: The AbstractProtocol should be protected member, which make it available for sub classes. And we need to use `application_protocol_.PackFrame` or `application_protocol_.UnpackFrame` when calling `Set` and `Get`.

File: common/include/serial_port.h [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+	boost::system::error_code ec_;
+
+	int 				baud_rate_;
```
Comment (luxianzi) [inline common/include/serial_port.h]: No need to use these variables, values are all in the config map.

File: common/source/serial_port.cpp [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+ *      Author: duanning
+ */
+
+#include "serial_port.h"
+
+SerialPort::SerialPort(void):
+	serial_port_(io_s_) {
+}
+
+error_condition SerialPort::Init(const string& name,\
```
Comment (luxianzi) [inline common/source/serial_port.cpp]: Please consider using `ReConfig` instead of this function.

File: common/source/serial_port.cpp [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+	character_size_ = character_size;
+	parity_			= parity;
+	stop_bits_		= stop_bits;
+	flow_control_	= flow_control;
+	if (ReConfig(config) != kNoError)
+		return make_error_condition(errc::io_error);
+	return kNoError;
+}
+
+error_condition SerialPort::Init(const string& name, const AbstractProtocol& protocol,\
```
Comment (luxianzi) [inline common/source/serial_port.cpp]: Same as above.

File: common/source/serial_port.cpp (hunk 12 above)
Comment (luxianzi) [reply common/source/serial_port.cpp]: Please store the config in map.

File: common/source/serial_port.cpp [hunk 21]
```diff
@@ -0,0 +1,132 @@
...
+		serial_port_.set_option(serial_port::flow_control(serial_port::flow_control::none));
+	else if (flow_control_ == flow_control_software)
+		serial_port_.set_option(serial_port::flow_control(serial_port::flow_control::software));
+	else
+		serial_port_.set_option(serial_port::flow_control(serial_port::flow_control::hardware));
+	SetInitialized();
+	return kNoError;
+}
+
+error_condition SerialPort::DeConfig() {
```
Comment (luxianzi) [inline common/source/serial_port.cpp:112]: Do we need to check if we are initialized?

File: common/source/serial_port.cpp [hunk 22]
```diff
@@ -0,0 +1,132 @@
...
+	serial_port_.close();
+	return kNoError;
+}
+
+error_condition SerialPort::Set(const string& key, const unsigned int value) {
+
```
Comment (luxianzi) [inline common/source/serial_port.cpp:123]: Please consider we are using `SerialPort::Set("baudrate". 115200)` to change the configuration map in other object which is referencing this class.

File: common/source/serial_port.cpp [hunk 23]
```diff
@@ -0,0 +1,132 @@
...
+	if (key == kSerialPortData) {
+		if (serial_port_.write_some(buffer(value), ec_) == 0)
+			return make_error_condition(errc::io_error);
+	}
+	return kNoError;
+}
+
+error_condition SerialPort::Get(const string& key, unsigned int& result) {
+	if (!IsInitialized())
```
Comment (luxianzi) [inline common/source/serial_port.cpp:148]: Also we get information like `SerialPort::Get("baudrate". baudrate)` in other object which is referencing this class.

File: common/include/serial_port.h [hunk 24]
```diff
@@ -0,0 +1,70 @@
...
+				const vector<uint8_t>& value);
+	error_condition Get(const string& key, unsigned int& result);
+	error_condition Get(const string& key, \
+				vector<uint8_t>& result);
+
+private:
+	io_service io_s_;
+	serial_port serial_port_;
+	AbstractProtocol application_protocol_;
+	boost::system::error_code ec_;
```
Comment (luxianzi) [inline common/include/serial_port.h]: Could we just make the `ec_` to be a local variable? Defined in the functions which need it.

**Review by luxianzi — DISMISSED** [review]

File: common/include/serial_port.h (hunk 4 above)
Comment (BitBander) [reply common/include/serial_port.h]: Okay

File: common/source/serial_port.cpp (hunk 12 above)
Comment (BitBander) [reply common/source/serial_port.cpp]: I have already did it

File: common/source/serial_port.cpp (hunk 11 above)
Comment (BitBander) [reply common/source/serial_port.cpp:15]: It will be closed

File: common/include/dummy_protocol.h [hunk 25]
```diff
@@ -0,0 +1,32 @@
...
+#include <iostream>
+#include "abstract_protocol.h"
+
+class DummyProtocol: public AbstractProtocol {
+public:
+	DummyProtocol() {};
+	~DummyProtocol() {};
+
+	virtual error_condition PackFrame(\
+			const vector<uint8_t>& buffer, vector<uint8_t>& out_data) {};
```
Comment (luxianzi) [inline common/include/dummy_protocol.h]: In this dummy protocol, `out_data` should the same as `buffer`.

File: common/include/dummy_protocol.h [hunk 26]
```diff
@@ -0,0 +1,32 @@
...
+	virtual error_condition UnpackFrame(\
+			vector<uint8_t>& buffer, vector<uint8_t>& valid_data) {};
```
Comment (luxianzi) [inline common/include/dummy_protocol.h]: `valid_data` needs to be the same as `buffer`.

File: common/include/dummy_protocol.h [hunk 27]
```diff
@@ -0,0 +1,32 @@
...
+private:
+	virtual error_condition IsFrameComplete(const vector<uint8_t>& buffer) {};
```
Comment (luxianzi) [inline common/include/dummy_protocol.h]: These functions needs to return a bool value, and we need to always return `true`.

File: common/include/dummy_protocol.h [hunk 28]
```diff
@@ -0,0 +1,32 @@
...
+	virtual error_condition IsFrameValid(const vector<uint8_t>& buffer) {};
```
Comment (luxianzi) [inline common/include/dummy_protocol.h]: Same as above

**Comment by luxianzi** [comment]
Still need to improve, let me do that.

**Pull request closed without merging**